Update configuring docs (#51)
fatkodima/online_migrations: docs/configuring.md

and a lock timeout for migrations can be configured via the `lock_retrier`.

> **Note**: If you do not set a lock retrier the default `NullLockRetrier` is used and it has a `lock_timeout` value of `0` that [disables the timeout](https://www.postgresql.org/docs/16/runtime-config-client.html#GUC-LOCK-TIMEOUT).

Reviewer: It does not disable the timeout, because that would be a 💥 
If the lock retrier is set to `nil`, then no lock retries are used. It is set in the generated initializer file when run `rails online_migrations:install`.

I think, this line should be removed. From the text above it is clear, that lock retries are used only when it is configured. And on line 106 is described that we can set it to `nil`.

Author: Thanks for clarifying. I see the default in the generator is in fact the [ExponentialLockRetrier](https://github.com/fatkodima/online_migrations/blob/9cebfd00b870ab246e813d8109767ccaf10a7df6/lib/generators/online_migrations/templates/initializer.rb.tt#L59C43-L59C65) so the timeout is setup by default and my initial assumption is wrong. 

As you note in the [changelog](https://github.com/fatkodima/online_migrations/blob/9cebfd00b870ab246e813d8109767ccaf10a7df6/CHANGELOG.md?plain=1#L46), if folks do not set a `lock_timeout` value for their LockRetrier implementation, the their retrier behaviour will not set a `lock_timeout` value and it will [fallback to the default](https://github.com/fatkodima/online_migrations/blob/9cebfd00b870ab246e813d8109767ccaf10a7df6/lib/online_migrations/lock_retrier.rb#L91-L93). This fallback may result in their migration running without a lock timeout depending how their server or connection is configured via `config/database.yml`, e.g. they don't have a `lock_timeout` configured in either. 

I'll note that when i first chose to use this library I wasn't sure about the lock retry behaviours so i set `config.lock_retrier = nil`. This config setting means the `NullLockRetrier` is [used as a fallback](https://github.com/fatkodima/online_migrations/blob/9cebfd00b870ab246e813d8109767ccaf10a7df6/lib/online_migrations/config.rb#L219) and that meant our migrations were running without a lock_timeout as we don't explictly set one in the db config and server ran with the pg default (0) which disables the timeout.

I think it's worth noting here in the docs that there are caveats around the choice of lock retrier implementation and how the `lock_timeout` is configured. As you say running without a `lock_timeout` can result in 💥 

Would you be open to me adding some clarity on these points in a follow up PR?

Author: ```suggestion
```

Reviewer: > Would you be open to me adding some clarity on these points in a follow up PR?

Yeah, anything that can help future users of this library. I think you can even update this PR.

Author: attempted in 74d0b24e1f455103fd24ea5ff1177da5dab62149

OOI - should the [NullLockRetrier](https://github.com/fatkodima/online_migrations/blob/9cebfd00b870ab246e813d8109767ccaf10a7df6/lib/online_migrations/lock_retrier.rb#L232) have a `lock_timeout` value of 0? If you want to ensure the fallback behaviour to the `config/database.yml` or server setting then this should be nil, right? 

Happy to PR this if it helps.

Reviewer: Updated `NullLockRetrier#lock_timeout` method to not be confusing (commit on `master`). That value is even not used, because of the `with_lock_retries` implementation, which just `yield`s without doing anything.

Author: 🙏🏼 thanks for clarifying my understanding and the code update in f70f166e509ae19fc78c5e41d3c22948e71670b5

)
```

> When statement within transaction fails - the whole transaction is retried. Similarly if any statement fails when running outside a transaction then the whole migration is retried, e.g. using `disable_ddl_transaction!`.

Reviewer: ```suggestion
When statement within transaction fails - the whole transaction is retried. If any statement fails when running outside a transaction (e.g. using `disable_ddl_transaction!`) then only that statement is retried.
```
